### DISCUSSION TITLE
Wake model

### DISCUSSION THREAD

**Discussion by prith-gs** [NatLabRockies/floris#547] [Q&A] [answered]
Can I add a new (custom) wake deficit model on my own and use it to calculate AEP for a wind farm?

**Comment by rafmudaf** [comment, marked as answer]
Absolutely! We don't have formal documentation for this just yet so I'll try to describe it fully here. A good starting point is to copy a model that is similar to the one you'll be implementing and modify it.

You can add a new model class in the [wake_velocity](https://github.com/NREL/floris/tree/main/floris/simulation/wake_velocity) directory and register it with the framework by listing it in [Wake.model_map](https://github.com/NREL/floris/blob/main/floris/simulation/wake.py#L50). Also, be sure to include it in the wake_velocity directory's [`__init__.py`](https://github.com/NREL/floris/blob/main/floris/simulation/wake_velocity/__init__.py).

The class definition itself must have a couple of functions. First, there's a function to extract any information from the grid and flow field for sizing the data arrays. This information should be available at init and should not change throughout the calculation. If there is a bit of data that will change, pass it in the next function and not here.

```python
def prepare_function(self, grid: Grid, flow_field: FlowField) -> Dict[str, Any]
```

Next is the function to do the wake deficit calculation. Everything before the `*` is given to all of the wake deficit models. The input arguments listed after the `*` should match what is returned from `prepare_function` above.

```python
    def function(
        self,
        x_i: np.ndarray,
        y_i: np.ndarray,
        z_i: np.ndarray,
        axial_induction_i: np.ndarray,
        deflection_field_i: np.ndarray,
        yaw_angle_i: np.ndarray,
        turbulence_intensity_i: np.ndarray,
        ct_i: np.ndarray,
        hub_height_i,
        rotor_diameter_i,
        *,
        variables_from_prepare_function
    ) -> None:
```

Some models require a special grid and/or solver, and that mapping happens in [floris.simulation.Floris](https://github.com/NREL/floris/blob/main/floris/simulation/floris.py#L145). Modify this as needed for your model.

This should get you started, but I'm sure I'm missing something, so please update here with your progress and I'll keep helping however I can.

**Reply by prith-gs** [reply]
Hi @rafmudaf! Thanks for your help! I've started working on implementing a new model and I have a question on the dimensions of Ct.
In TurboPark, I see in line 83, one of the inputs is Cts which is believe is a 5-dimensional array.
![image](https://user-images.githubusercontent.com/103917694/224224503-688a6c26-44e9-40cd-bada-8f0b7eb72fd0.png)

In the Cumulative curl model,  I see in line 80, one of the inputs is ct which is believe is a 3-dimensional array.
![image](https://user-images.githubusercontent.com/103917694/224224051-a3e45788-b43c-41c5-8bd6-be9827b29337.png)

What are the different dimensions in Cts and ct mean? 

Also, I'm trying to implement an iterative model similar to the Ainslie wake model. Do you have any recommendations on how to implement such a model?

**Reply by Darry6682** [reply]
> Absolutely! We don't have formal documentation for this just yet so I'll try to describe it fully here. A good starting point is to copy a model that is similar to the one you'll be implementing and modify it.
> 
> You can add a new model class in the [wake_velocity](https://github.com/NREL/floris/tree/main/floris/simulation/wake_velocity) directory and register it with the framework by listing it in [Wake.model_map](https://github.com/NREL/floris/blob/main/floris/simulation/wake.py#L50). Also, be sure to include it in the wake_velocity directory's [`__init__.py`](https://github.com/NREL/floris/blob/main/floris/simulation/wake_velocity/__init__.py).
> 
> The class definition itself must have a couple of functions. First, there's a function to extract any information from the grid and flow field for sizing the data arrays. This information should be available at init and should not change throughout the calculation. If there is a bit of data that will change, pass it in the next function and not here.
> 
> ```python
> def prepare_function(self, grid: Grid, flow_field: FlowField) -> Dict[str, Any]
> ```
> 
> Next is the function to do the wake deficit calculation. Everything before the `*` is given to all of the wake deficit models. The input arguments listed after the `*` should match what is returned from `prepare_function` above.
> 
> ```python
>     def function(
>         self,
>         x_i: np.ndarray,
>         y_i: np.ndarray,
>         z_i: np.ndarray,
>         axial_induction_i: np.ndarray,
>         deflection_field_i: np.ndarray,
>         yaw_angle_i: np.ndarray,
>         turbulence_intensity_i: np.ndarray,
>         ct_i: np.ndarray,
>         hub_height_i,
>         rotor_diameter_i,
>         *,
>         variables_from_prepare_function
>     ) -> None:
> ```
> 
> Some models require a special grid and/or solver, and that mapping happens in [floris.simulation.Floris](https://github.com/NREL/floris/blob/main/floris/simulation/floris.py#L145). Modify this as needed for your model.
> 
> This should get you started, but I'm sure I'm missing something, so please update here with your progress and I'll keep helping however I can.

Hi @rafmudaf  I am currently working on incorporating my own wake superposition model into FLORIS for wake steering within wind farms. My model integrates both a wake model and a wake deflection model, and it has been published in Energy and Renewable Energy (https://doi.org/10.1016/j.energy.2023.128402   https://doi.org/10.1016/j.renene.2024.120324).

I have primarily used MATLAB for my simulations and am relatively new to Python, which makes this integration challenging for me. I would greatly appreciate any guidance or documentation that could help me include my custom wake model in FLORIS. Specifically, I am looking for:

Steps to integrate a custom wake model into FLORIS.
Any examples or templates that could serve as a reference.
Best practices for ensuring compatibility and performance within the FLORIS framework.
Thank you very much for your assistance. Your help will be invaluable in advancing my research.

**Reply by rafmudaf** [reply]
Hi @Darry6682 the comment above is probably still the best documentation we have of adding a wake mode. I also added some info at https://nrel.github.io/floris/dev_guide.html#extending-the-models. I suggest to start by copying an existing model (good ones for simplicity are Jensen and Jimenez for velocity and deflection), change their names and connect them as described above, and then verify that they work. Maybe change something obvious in the model to ensure that it's different from the one you copied. Then, start adding your model components from there.

I'm happy to provide some support as you do this. Please open a new discussion and describe what you intend to do in some detail (introduce the model from your paper, describe any details of how it's solved). Then, as you do it, you can update your discussion and I can follow along. If you have any questions, we can work through them in a thread there.

**Comment by prith-gs** [comment]
Thanks so much @rafmudaf! This is some good info for me to start working on this. I'll come back to this discussion with further questions as I make progress.

**Reply by FulaiJin** [reply]
Hello, I also plan to write the new wake model into floris. What content should I modify?

**Reply by rafmudaf** [reply]
Hi @FulaiJin Thanks for letting us know you're planning to implement a new model. Can you describe what you've done so far and give more detail on your questions?

**Reply by rafmudaf** [reply]
In fact @FulaiJin, it might be best to open a new discussion thread with the details of your planned model and describing your process as you. To start, it can be as simple as "I've read this, that, and this other GitHub Issues and Discussions. My plan is to now do x, y, z." If you start that thread, I can easily work with you to design your implementation and work through any details.

**Reply by FulaiJin** [reply]
Thanks rafmudaf for your reply, I will take a while to sort out my work and open a new discussion thread.
